include/netdb.h

```diff
@@ -184,14 +184,17 @@ struct hostent
   FAR char **h_aliases;    /* A pointer to an array of pointers to the
                             * alternative host names, terminated by a
                             * null pointer. */
-  int        h_addrtype;   /* Address type. */
-  int        h_length;     /* The length, in bytes, of the address. */
+  FAR int   *h_addrtypes;  /* A pointer to an array of address type. */
+  FAR int   *h_lengths;    /* A pointer to an array of the length, in bytes,
+                            * of the address. */
   FAR char **h_addr_list;  /* A pointer to an array of pointers to network
                             * addresses (in network byte order) for the host,
                             * terminated by a null pointer. */
 };
 
 #define h_addr h_addr_list[0] /* For backward compatibility */
+#define h_length h_lengths[0]
+#define h_addrtype h_addrtypes[0]
```

Comment:
i'm afraid that changing a structure which is a part of standard api this way.

Reply:
> i'm afraid that changing a structure which is a part of standard api this way.

It's hard to return both IPv4 and IPv6 in one hostent, since hostent has only one h_addrtype/h_length field. The modification extend h_addrtype/h_length to array by following h_addr_list/h_addr design pattern, and provide macro to keep the backward  compatibility. The most caller work fine with the new struct and macro.

To keep the compatibility, caller shouldn't directly touch  h_lengths/h_addrtypes fields, they must call the new getaddrinfo/freeaddrinfo API instead, like this:
https://github.com/apache/incubator-nuttx-apps/pull/151
https://github.com/apache/incubator-nuttx-apps/pull/153

The array extension is used internally to support getaddrinfo:
https://github.com/apache/incubator-nuttx/pull/664/files#diff-cf02c526db0b2c6580f921a4083b5ce8R290

This method is simplest approach with some backword compatibility what I can think.

Reply:
as gethostbyname etc should not return hostent containing multiple types anyway, there's little point to change struct hostent.
for getaddrinfo, it's better to implement gethostbyname2 and use it.

Reply:
> as gethostbyname etc should not return hostent containing multiple types anyway, there's little point to change struct hostent.
> for getaddrinfo, it's better to implement gethostbyname2 and use it.

I also consider this solution before but:
1.getaddrinfo need call gethostbyname2 twice.
2.dns cache hold the name to address(multiple entries) mapping. It's hard to manage the dns cache if we split IPv4/IPv6 query into two call.

Reply:
i don't see what's wrong with calling gethostbyname2 twice. it's certainly better than breaking gethostbyname.

Reply:
> i don't see what's wrong with calling gethostbyname2 twice. it's certainly better than breaking gethostbyname.

But gethostbyname isn't broken, many caller in apps still work as before without any modification. The beahviour is also same as before.

Reply:
if h_addr_list contains both of ipv4 and ipv6 addresses, an app expecting the standard behavior (thus only look at h_length/h_addrtype) would break.

Reply:
> if h_addr_list contains both of ipv4 and ipv6 addresses, an app expecting the standard behavior (thus only look at h_length/h_addrtype) would break.

Yes, if the caller touch the non first address which has the different type from the first one. But the problem already exist there even without my modification:
https://github.com/apache/incubator-nuttx/pull/664/files#diff-b38c46bf06ff1cf66969bedcd951db60L404
https://github.com/apache/incubator-nuttx/pull/664/files#diff-b38c46bf06ff1cf66969bedcd951db60L564
 
My patch doesn't introduce the new problem but just extend h_length/h_addrtype to express the fact(IPv4/IPv6 combination) correctly. If this case is a concern need to fix. It's better to add a patch on top of mine to remove the different IP address from the list.

Reply:
IMO,
* if it's already the case, it should be fixed.
* it isn't a good idea to rely on the bug. especially when it involves standard violation.

if you need a libc-internal non-standard gethostbyname-like interface, why don't you just introduce it with another name?

Reply:
Ok, I will add an internal struct and funciton to return the full list.